src/node_dir.h

```diff
@@ -0,0 +1,87 @@
+#ifndef SRC_NODE_DIR_H_
+#define SRC_NODE_DIR_H_
+
+#if defined(NODE_WANT_INTERNALS) && NODE_WANT_INTERNALS
+
+#include "node_file.h"
+#include "node.h"
+#include "req_wrap-inl.h"
+
+namespace node {
+
+using v8::Context;
+using v8::FunctionCallbackInfo;
+using v8::HandleScope;
+using v8::Local;
+using v8::MaybeLocal;
+using v8::Null;
+using v8::Object;
+using v8::Persistent;
+using v8::Promise;
+using v8::Value;
+
+using fs::FSReqBase;
+using fs::FSReqAfterScope;
+using fs::FSReqWrapSync;
+
+namespace fs_dir {
+
+class DirHandle;
+class DirReq;
+
+// XXX Needed to have access to the DirHandle in async opendir so that we can
+// set the uv_dir_t.
+class DirReq : public FSReqBase {
+ public:
+  static DirReq* from_req(uv_fs_t* req) {
+    return static_cast<DirReq*>(ReqWrap::from_req(req));
+  }
+
+  DirHandle* handle_;
+};
+
+// Needed to propagate `uv_dir_t`.
+class DirHandle : public AsyncWrap {
+ public:
+  static constexpr int kDirHandleFieldCount = 1;
+
+  static DirHandle* New(Environment* env, uv_dir_t* dir);
+  ~DirHandle() override;
+
+  static void New(const v8::FunctionCallbackInfo<v8::Value>& args);
+  static void Open(const FunctionCallbackInfo<Value>& args);
+  static void Read(const FunctionCallbackInfo<Value>& args);
+  static void Close(const FunctionCallbackInfo<Value>& args);
+
+  inline uv_dir_t* dir() { return dir_; }
+  AsyncWrap* GetAsyncWrap() { return this; }
+
+  void MemoryInfo(MemoryTracker* tracker) const override {
+  }
```

Comment:
This should at least track `dir_`, I think, because it owns that object? i.e.

```suggestion
    if (dir_ != nullptr) tracker->TrackFieldWithSize("dir", sizeof(*dir_));
  }
```

(And then unset `dir_` to `nullptr` after the close operations. If that seems unnecessary, you can also just remove the conditional, I guess.)